### DISCUSSION TITLE
Difficulty Visualizing Navigation Regions Created with Code

### DISCUSSION THREAD

**Discussion by Booblesnoot42** [godotengine/godot-proposals#9641] [2D]
I've been trying to get a handle on creating navigation regions using code rather than using the NavigationRegion2D node. I was planning on making a small test scene and using the navigation debugger to see the regions and make sure they were how I expected them to be, but after a few days of trying without being able to get anything to appear I've realized that the debugger only shows regions made using the node. I know the region exists because my test agent can use it to navigate, but it doesn't get drawn.

My next plan was to get the polygons from the regions using NavigationServer2D and draw them from that, but as far as I can tell there's no way to get the polygon of the region; there is a setter function (NavigationServer2D.region_set_navigation_polygon(...)), but no getter.

Unless I'm missing something, I'd like to suggest that either or both of these gets changed; I'm assuming the code-based regions not being drawn is an oversight, and I'm not sure why a user shouldn't be able to get the polygons of the regions. My next plan for debugging is to just keep a copy of all the navigation polygons to draw, but I'd rather not have to do that.

**Comment by smix8** [comment]
The navigation debug visuals are currently all part of the navigation related nodes so there are no debug visuals for anything that just exists inside the NavigationServer without a Node in the SceneTree.

There are plans to change this, see proposal https://github.com/godotengine/godot-proposals/issues/7208.

If you want to draw your own debug for a NavigationPolygon resource it is a matter of drawing polygons for each polygon.

1. get the NavigationPolygon.get_polygon_count()
2. loop over the polygon count range and get each NavigationPolygon.get_polygon(index).
3. the array that you get are the indices that make up the polygon. Use the indices on the NavigationPolygon.get_vertices() array to get the Vector2 positions for the polygon.
4. Now that you have the positions use the Node2D draw function or the RenderingServer canvas_item API to add e.g. polygons or polylines. If your region has a transform you need to apply it to each vertex position so that the visuals end up in the correct place.

That the NavigationServer even allows to set a resource directly on the API is legacy mold and bound to be removed. See proposal https://github.com/godotengine/godot-proposals/issues/6317.

A server runs with RID handles, allowing users to set a resource directly creates a ton of issues on top of needing multiple data copies for nothing. If you update something with a server API only it is assumed that you have a handle to whatever you are updating and a handle to whatever data you are using for the upate. The NavigationPolygon is basically your handle for the data, don't throw away your handles.
